smarts/core/renderer.py

```diff
@@ -0,0 +1,261 @@
+# Copyright (C) 2020. Huawei Technologies Co., Ltd. All rights reserved.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+
+import os
+import logging
+from typing import NamedTuple
+import importlib.resources as pkg_resources
+
+import gltf
+from direct.showbase.ShowBase import ShowBase
+from panda3d.core import (
+    ClockObject,
+    NodePath,
+    Shader,
+    loadPrcFileData,
+    FrameBufferProperties,
+    GraphicsOutput,
+    GraphicsPipe,
+    OrthographicLens,
+    Texture,
+    WindowProperties,
+)
+
+from . import glsl, models
+from .masks import RenderMasks
+from .colors import SceneColors
+from .scenario import Scenario
+from .coordinates import Pose
+
+
+class _ShowBaseInstance(ShowBase):
+    """ Wraps a singleton instance of ShowBase from Panda3D. """
+
+    def __new__(cls, timestep_sec=0.1):
+        # Singleton pattern:  ensure only 1 Renderer
+        if "__it__" not in cls.__dict__:
+            # disable vsync otherwise we are limited to refresh-rate of screen
+            loadPrcFileData("", "sync-video false")
+            loadPrcFileData("", "model-path %s" % os.getcwd())
+            loadPrcFileData("", "audio-library-name null")
+            loadPrcFileData("", "gl-version 3 3")
+            loadPrcFileData("", "notify-level error")
+            loadPrcFileData("", "print-pipe-types false")
+            # https://www.panda3d.org/manual/?title=Multithreaded_Render_Pipeline
+            # loadPrcFileData('', 'threading-model Cull/Draw')
+            # have makeTextureBuffer create a visible window
+            # loadPrcFileData('', 'show-buffers true')
+        it = cls.__dict__.get("__it__")
+        if it is None:
+            cls.__it__ = it = object.__new__(cls)
+            it.init(timestep_sec)
+        else:
+            assert timestep_sec == it._timestep_sec
```

Comment:
question: for the subsequent times after the first time we create `_ShowBaseInstance`, do we allow users to create _ShowBaseInstance (or create smarts, since creating smarts creates _ShowBaseInstance) again with a different initial `timestep_sec`?

Reply:
If we don't allow that, maybe it is good add a message to let the user know

```suggestion
            assert timestep_sec == it._timestep_sec, "Running multiple instances of SMARTS must use same timestep second"
```

Reply:
Agree.  Will add the message.
(BTW, as far as I know, we don't have any circumstances where multiple SMARTS instances are created simultaneously, which is where this would be an issue.  Back-to-back, but not overlapping usage of `_ShowBaseInstance` with different values for `timestep_sec` should be fine.)